### ISSUE TITLE
crash in library wrapper

### ISSUE THREAD

**Issue by allisonkarlitskaya** [allisonkarlitskaya/systemd_ctypes#63]
This just randomly happened while running cockpit's tests.  Never saw it before.

Line 72 is:

```python
        # workaround https://github.com/systemd/systemd/issues/23826
        # keep calling wait() until there's nothing left
        while self.sd_event.wait(0):
            self.sd_event.dispatch()                 <--- this one
            while self.sd_event.prepare():
                self.sd_event.dispatch()
```

Line 62 is:

```python
def instancemethod(func):
    func.argtypes = [ctypes.c_void_p, *func.argtypes]

    def wrapper(*args):
        return func(*args)            <--- this one
    return wrapper
```

test/pytest/test_bridge.py::test_channel[HttpChannel] Fatal Python error: Aborted

Thread 0x00007f04c59596c0 (most recent call first):
  File "/usr/lib64/python3.12/threading.py", line 299 in __enter__
  File "/usr/lib64/python3.12/threading.py", line 524 in release
  File "/usr/lib64/python3.12/concurrent/futures/thread.py", line 87 in _worker
  File "/usr/lib64/python3.12/threading.py", line 1010 in run
  File "/usr/lib64/python3.12/threading.py", line 1073 in _bootstrap_inner
  File "/usr/lib64/python3.12/threading.py", line 1030 in _bootstrap

Current thread 0x00007f04c7d13740 (most recent call first):
  File "/var/home/lis/src/cockpit/fsinfo/src/cockpit/_vendor/systemd_ctypes/librarywrapper.py", line 62 in wrapper
  File "/var/home/lis/src/cockpit/fsinfo/src/cockpit/_vendor/systemd_ctypes/event.py", line 72 in select
  File "/usr/lib64/python3.12/asyncio/base_events.py", line 1933 in _run_once
  File "/usr/lib64/python3.12/asyncio/base_events.py", line 638 in run_forever
  File "/usr/lib64/python3.12/asyncio/base_events.py", line 671 in run_until_complete
  File "/usr/lib/python3.12/site-packages/pytest_asyncio/plugin.py", line 525 in inner
  File "/usr/lib/python3.12/site-packages/_pytest/python.py", line 194 in pytest_pyfunc_call
  File "/usr/lib/python3.12/site-packages/pluggy/_callers.py", line 80 in _multicall
  File "/usr/lib/python3.12/site-packages/pluggy/_manager.py", line 112 in _hookexec
  File "/usr/lib/python3.12/site-packages/pluggy/_hooks.py", line 433 in __call__
  File "/usr/lib/python3.12/site-packages/_pytest/python.py", line 1799 in runtest
  File "/usr/lib/python3.12/site-packages/_pytest/runner.py", line 169 in pytest_runtest_call
  File "/usr/lib/python3.12/site-packages/pluggy/_callers.py", line 80 in _multicall
  File "/usr/lib/python3.12/site-packages/pluggy/_manager.py", line 112 in _hookexec
  File "/usr/lib/python3.12/site-packages/pluggy/_hooks.py", line 433 in __call__
  File "/usr/lib/python3.12/site-packages/_pytest/runner.py", line 262 in <lambda>
  File "/usr/lib/python3.12/site-packages/_pytest/runner.py", line 341 in from_call
  File "/usr/lib/python3.12/site-packages/_pytest/runner.py", line 261 in call_runtest_hook
  File "/usr/lib/python3.12/site-packages/_pytest/runner.py", line 222 in call_and_report
  File "/usr/lib/python3.12/site-packages/_pytest/runner.py", line 133 in runtestprotocol
  File "/usr/lib/python3.12/site-packages/_pytest/runner.py", line 114 in pytest_runtest_protocol
  File "/usr/lib/python3.12/site-packages/pluggy/_callers.py", line 80 in _multicall
  File "/usr/lib/python3.12/site-packages/pluggy/_manager.py", line 112 in _hookexec
  File "/usr/lib/python3.12/site-packages/pluggy/_hooks.py", line 433 in __call__
  File "/usr/lib/python3.12/site-packages/_pytest/main.py", line 348 in pytest_runtestloop
  File "/usr/lib/python3.12/site-packages/pluggy/_callers.py", line 80 in _multicall
  File "/usr/lib/python3.12/site-packages/pluggy/_manager.py", line 112 in _hookexec
  File "/usr/lib/python3.12/site-packages/pluggy/_hooks.py", line 433 in __call__
  File "/usr/lib/python3.12/site-packages/_pytest/main.py", line 323 in _main
  File "/usr/lib/python3.12/site-packages/_pytest/main.py", line 269 in wrap_session
  File "/usr/lib/python3.12/site-packages/_pytest/main.py", line 316 in pytest_cmdline_main
  File "/usr/lib/python3.12/site-packages/pluggy/_callers.py", line 80 in _multicall
  File "/usr/lib/python3.12/site-packages/pluggy/_manager.py", line 112 in _hookexec
  File "/usr/lib/python3.12/site-packages/pluggy/_hooks.py", line 433 in __call__
  File "/usr/lib/python3.12/site-packages/_pytest/config/__init__.py", line 166 in main
  File "/usr/lib/python3.12/site-packages/_pytest/config/__init__.py", line 189 in console_main
  File "/usr/bin/pytest", line 8 in <module>
Aborted (core dumped)

**Comment by martinpitt** [comment]
Pretty much the same stack trace happened in yesterday's Debian builds of release 310, where it crashed that way on [riscv64](https://buildd.debian.org/status/fetch.php?pkg=cockpit&arch=riscv64&ver=310-1&stamp=1706724161&raw=0), [ppc64el](https://buildd.debian.org/status/fetch.php?pkg=cockpit&arch=ppc64el&ver=310-1&stamp=1706723947&raw=0), and [ppc64](https://buildd.debian.org/status/fetch.php?pkg=cockpit&arch=ppc64&ver=310-1&stamp=1706724036&raw=0).

However, there it crashed in `test_fsinfo_watch_identity_changes()`. Do you still have a log URL for the crash you saw? It seems much more plausible to me that this happens in the inotify test than in test_channel[HttpChannel] -- that doesn't use ctypes/inotify at all. Perhaps just log output stderr/out ordering issue?

**Comment by martinpitt** [comment]
This reproduces fairly reliably  on a Debian s390x porter box, with `pytest -k test_fsinfo_watch_identity_changes`. I don't even have to build the Debian package, cloning upstream and vendor/checkout are sufficient. The stack trace is not super useful, but points to calling `sd_event_dispatch()` in the wrong way?
```
Core was generated by `/usr/bin/python3 /usr/bin/pytest -k test_fsinfo_watch_identity_changes'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  __pthread_kill_implementation (threadid=<optimized out>, signo=<optimized out>, no_tid=no_tid@entry=0) at pthread_kill.c:44
44      pthread_kill.c: No such file or directory.
(gdb) bt
#0  __pthread_kill_implementation (threadid=<optimized out>, signo=<optimized out>, no_tid=no_tid@entry=0) at pthread_kill.c:44
#1  0x000003ff9ed176aa in __pthread_kill_internal (signo=<optimized out>, threadid=<optimized out>) at pthread_kill.c:78
#2  0x000003ff9ecc7e88 in __GI_raise (sig=<optimized out>) at ../sysdeps/posix/raise.c:26
#3  <signal handler called>
#4  0x000003ff9d1868fa in ?? () from /lib/s390x-linux-gnu/libffi.so.8
#5  0x000003ff9d186d00 in ?? () from /lib/s390x-linux-gnu/libffi.so.8
#6  0x000003ff9d1045f2 in ?? () from /lib/s390x-linux-gnu/libsystemd.so.0
#7  0x000003ff9d10483a in sd_event_dispatch () from /lib/s390x-linux-gnu/libsystemd.so.0
#8  0x000003ff9d186c58 in ?? () from /lib/s390x-linux-gnu/libffi.so.8
#9  0x000003ff9d1862ac in ?? () from /lib/s390x-linux-gnu/libffi.so.8
#10 0x000003ff9d18674e in ffi_call () from /lib/s390x-linux-gnu/libffi.so.8
#11 0x000003ff9d2938be in ?? () from /usr/lib/python3.11/lib-dynload/_ctypes.cpython-311-s390x-linux-gnu.so
#12 0x000003ff9d291a32 in ?? () from /usr/lib/python3.11/lib-dynload/_ctypes.cpython-311-s390x-linux-gnu.so
#13 0x0000000001107a84 in PyObject_Call ()
```

This at least matches the Py stack trace where it crashes:

```py
        while self.sd_event.wait(0):
            self.sd_event.dispatch()
            while self.sd_event.prepare():
                self.sd_event.dispatch()   # ← CRASHES HERE
```
The comment says this whole thing is a workaround for https://github.com/systemd/systemd/issues/23826 which got fixed 1.5 years ago. I naïvely tried to drop this whole loop, and that breaks things.

I keep the schroot debugging session open, perhaps we can look into this today?
